Remove support for disabling OTF reprojection

nyalldawson opened the pull request:
Because:

- OTF reprojection is mature and stable
- Should be no cost when not required - transforms are shortcut
when not required
- Reduces code complexity
- Canvas OTF support was being incorrectly used as a flag for
whether measurements should be made in ellipsoidal mode. Instead
the project's ellipsoid setting should be checked for this.
- Should simplify behavior for new users

nyalldawson committed:
Remove support for disabling OTF reprojection

Because
- OTF reprojection is mature and stable
- Should be no cost when not required - transforms are shortcut
when not required
- Reduces code complexity
- Canvas OTF support was being incorrectly used as a flag for
whether measurements should be made in ellipsoidal mode. Instead
the project's ellipsoid setting should be checked for this.
- Should simplify behavior for new users

nirvn commented:
A net -475 lines of code to maintain? Wouhou!

nyalldawson committed:
Fix failing tests

nyalldawson committed:
Remove support for setting map units in QgsMapSettings, QgsMapCanvas and QgsDxfExport

In a post-OTF-disabled world these are no longer required. The
map units are dictated by the map units for the destination
CRS.

It's dangerous to separate these two properties as setting
the incorrect map units for a given CRS will result in
incorrect scale calculations.

nyalldawson committed:
Improve documentation for QgsMapCanvas and remove false QGIS 3 TODO

timlinux commented:
Good stuff thanks!

nyalldawson committed:
Correctly use QgsMapCanvas::freeze instead of setRenderFlag

nyalldawson commented:
@wonder-sk in the dev list discussion about this you stated that you want to see a concept of "crs-less" layers first. Is this still the case, or are you ok for this to be merged without crs-less layers?

wonder-sk commented on src/core/dxf/qgsdxfexport.cpp:
```diff
@@ -556,6 +555,11 @@ int QgsDxfExport::writeToFile( QIODevice* d, const QString& encoding )
 
+QgsUnitTypes::DistanceUnit QgsDxfExport::mapUnits() const
+{
+  return mMapUnits;
```
Maybe better to return mCrs.mapUnits() so we can get rid of mMapUnits?

wonder-sk reviewed (approved):
Nice cleanup!

OK for me to merge without support for layers with invalid CRS - that can be added later...

m-kuhn commented:
I remember that some people used OTF reprojection to check if CRSes match. @gioman

NathanW2 commented:
Can we do that in a better way?  In Roam I made a page which lists all the
layers and the CRS for each one so you can see if there is any differences.

On Tue, Feb 28, 2017 at 10:00 PM, Matthias Kuhn <notifications@github.com>
wrote:

> I remember that some people used OTF reprojection to check if CRSes match.
> @gioman <https://github.com/gioman>
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/qgis/QGIS/pull/4198#issuecomment-282981545>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAXS3BQmOIpnThuufCv1tgAfu4fZ9Pomks5rg-IkgaJpZM4MN8TU>
> .
>

nyalldawson commented:
> I remember that some people used OTF reprojection to check if CRSes match. 

Yeah, I don't see that as a strong use case. Better to show crs in the layer tree or something

m-kuhn commented:
> Yeah, I don't see that as a strong use case. Better to show crs in the layer tree or something

I think it was more for the case where you don't know the real CRS of a layer (because undefined, wrong CRS assigned...) and you want to overlay two layers without OTF to see if their CRSes match. But let's wait for @gioman (I think it was him) to discuss alternative approaches.

saberraz commented:
-1 for removing this option. Very useful to check if the layers with unknown/misinterpreted CRSs line up.
In fact as default my OTF is off. For new users it becomes a nightmare to use as in real world layers come with unknown projections.

nyalldawson commented:
I still don't see this as a strong argument for leaving this option. Use the new "projection finder" processing algorithm instead.

Please keep in mind that this isn't removing features for the sake of it - this setting results in lots of bugs and incorrect measurement issues.

saberraz commented:
@nyalldawson I don't want to sound like https://xkcd.com/1172/ but all the users (esp. new ones) find auto OTF very confusing.

Almost all the time I and most of people who I know use QGIS with OTF off and enable it only for viewing online background maps...and then turn off afterwards.

Talking about the processing toolbox...there is an ongoing old bug which still outputs raster layers in 4326 (tested in QGIS master...and still there...but this is a different topic). Things like this will be extremely confusing for users if OTF is not disabled.

nyalldawson commented:
I'm still dead against keeping this code... But what about this compromise:

We add a "no projection" choice in the project CRS selector. This would mean no projection takes place at all, and there's no CRS for the project/map. Stuff is just rendered where the raw coordinates fall. Measurements are total nonsense and might mean anything (much like with the current otf off option).

m-kuhn commented:
... or have a per-layer option "use project CRS instead"?

saberraz commented:
I am fine with no projection and a plain XY canvas....keeps CAD/Mapinfo users happy! :)

nyalldawson commented:
> or have a per-layer option "use project CRS" instead?

That's getting back to Martin's idea for layers with no CRS instead of the CRS selector dialog idea.

gioman commented:
Hi all,

> -1 for removing this option. Very useful to check if the layers with unknown/misinterpreted CRSs line up.
> In fact as default my OTF is off. For new users it becomes a nightmare to use as in real world layers come with unknown projections.

agree,

having the ability to disable OTR is useful not only from a didactic point of view, but also from a practical one. Countless are still the occasions where unlucky users have to figure about shapefiles without the .prj file (still a common thing out of Arcgis, which is also "able" to output filegeodatabases without CRSs) or geotiffs with just the world file and no metadata whatsoever. Frequently is not just a matter of choosing between a projected national CRS and WGS84 (which has an obvious solution): several are the cases were the possible options are between several projected CRSs where the only thing that changes is the origin coordinate (sometimes of just a few tens of meters) or the false origin parameter. I use OTF off/on also frequently to show the precision differences and implications of using NTV2 transformations or towgs84 parameters.


> We add a "no projection" choice in the project CRS selector. This would mean no projection takes place at all, and there's no CRS for the project/map. Stuff is just rendered where the raw coordinates fall. Measurements are total nonsense and might mean anything (much like with the current otf off option).

not sure I understand correctly the proposal, but seems like a solution, right?

m-kuhn commented:
... I think these also are all solvable by setting the layer in question to the "project CRS" (i.e. disabling reprojection for a particular layer).

gioman commented:
> ... I think these also are all solvable by setting the layer in question to the "project CRS" (i.e. disabling reprojection for a particular layer).

cool.

m-kuhn commented:
So we could apply Martin's proposal first, @gioman and @saberraz can review if this works well and in case we find trouble we can still apply the "no projection" option for the project CRS?

gioman commented:
> So we could apply Martin's proposal first, @gioman and @saberraz can review if this works well and in case we find troubles we can still apply the "no projection" option for the project CRS?

I will be happy to test the PR/patch.

nyalldawson commented:
> I think these also are all solvable by setting the layer in question to the "project CRS" (i.e. disabling reprojection for a particular layer).

Isn't that rather complex for users to understand? And I'd still see it as prone to mismeasurements. 

Having a "no CRS" choice (which is reflected by giving NO units for measurements, and no option to choose a unit in the measurement dialog) makes it explicitly clear that measurements in this case are more or less artbitrary.

I can also see a good use case for "no CRS" for people who use QGIS for non earth uses... Like rpg maps or building mapping or microscopic stuff (apparently ppl do this!)

nyalldawson commented:
Actually, thinking more about this, both options effectively come down to the same change: existing a "no CRS" option in the projection selector which returns an invalid CRS.

If either the layer or the map settings/canvas has an invalid CRS then the transform will be shortcut and return coordinates unchanged - which effectively addresses both options. Invalid canvas CRS = shortcut ALL transforms, invalid layer CRS = shortcut only that layer's transform.

saberraz commented:
For canvas: there should be a no projection option just an XY canvas with an unknown units. As @nyalldawson said all the measurements are reported as ???? and it is up to the user to interpolate the units.

For layers, in QGIS options there should be another option (default) to apply No CRS for layers with unknown/missing projections. Currently we have:

- Prompt for assigning a projection
- Project CRS
- Default CRS

With the above combinations, IMO, all use cases can be covered.

nyalldawson committed:
More efficient use of strings

nyalldawson committed:
Make QgsCoordinateReferenceSystem less noisy with debug messages

nyalldawson committed:
An invalid CRS should return unknown units for mapUnits

nyalldawson committed:
Allow a "no projection" option in projection selection widget/dialog

This option allows users to select "no projection", when
there's no suitable projection to select...

nyalldawson committed:
[FEATURE] Allow setting project CRS to "no projection"

This replaces the older "OTF off" setting. If a project has
no projection set, then the following things happen:

- no reprojection is done while rendering - features are just
drawn using their raw coordinates
- ellipsoid, distance and area units, and coordinate display
are locked out.
- all measurements are done in unknown map units, and no conversion
is possible.

As well as addressing the use cases in which users disabled
OTF reprojection, this setting may also be useful for users
who use QGIS to create "non-earth" maps, such as RPG or fantasy
maps, or for microscopic analysis, or some other local reference
system not related to earth based projections.

nyalldawson commented:
Ok - I've done the canvas/project part. You can now set a project to have "no projection", with the following consequences:

- no reprojection is done while rendering - features are just drawn using their raw coordinates
- ellipsoid, distance and area units, and coordinate display are locked out and forced to "map units" (ellipsoid is forced to none)
- all measurements are done in unknown map units, and no conversion is possible.

I can't spend any more time on this, but I've done the hard part and it should be quite trivial to now add the no-projection support for layers too. Someone would need to:

- make sure QgsProjectionSelectionWidget plays nice when setting `QgsProjectionSelectionWidget::dialog()->setShowNoProjection( true )`
- make sure the layer properties dialogs don't reject invalid crses
- make sure invalid CRSes are correctly saved/restored to layers inside projects
- add the option for unknown layers to default to project crs

nyalldawson committed:
Indentation

saberraz commented:
@nyalldawson where is the setting to set a project to No CRS?

saberraz commented:
Ahhhh...found it eventually! IMO, there should be an option for this.

![image](https://cloud.githubusercontent.com/assets/2649047/23465790/38bb41de-fe91-11e6-8625-4de703f6bfdb.png)

SrNetoChan commented:
Starting a project with "No projection", once you load a layer, the project's CRS is set to match the layer, but the bottom-right display is not updated. Bug?

![image](https://cloud.githubusercontent.com/assets/3607161/23467070/f0040080-fe94-11e6-8c42-24938e993b75.png)

 Not even if you open the Project properties CRS tab and click OK

![image](https://cloud.githubusercontent.com/assets/3607161/23467140/22857962-fe95-11e6-8982-e4a68d90d7fe.png)

haubourg commented:
Hi, 
2 days between PR and merge for such a major change puzzles me... Especially if we don't have enough time to polish that. We talk for months in QGIS API 3.0 changes for smaller things. 
Code is hard to maintain and caused many issues, that I can understand.
 In the other hand,  I have trained dozens of trainees to just check off OTF to see real XY stored in the dataset and I think we all did the same. Mapinfo and Arcgis both behaves that way too. So my point is that we need a very clear and advertised explanation for that change, their reasons, and the new tips to set the no projection option. So code is one thing, help users change is one we do need to keep in mind too. Maybe a warning tip in crs chooser dialog in qgis 3.0 and 3.2 could be the way to lead the attention of users directly to the new way of setting 'no projection'. 

A mail in the list and a ticket in QGIS 3 api is the least we can ask. Don't forget we try to coordinate efforts to build QGIS 3 together and such sudden unadvertised make that much more difficult.

nyalldawson commented:
@hauborg I understand your concerns, but in the end this change won't be significantly different for users.

I'll move the "no projection" option up from the tree view to its own widget to make it more prominent, and put it up near where the old OTF check box was. 

In the end the only really noticeable difference for users vs OTF-off is that measurements are all forced to be planar and unitless. (Which will hopefully finally put and end to the ongoing confusion about measurements in QGIS. Now, if you use a valid projection everything is perfect. If you use no projection, you get the only accurate measurements we can do (planar, no units)).

So really from an end users pov this is just a string change (unchecking "on the fly" vs checking "no projection") + addition of the feature that it's now easy to make maps using a non earth based projection.

It'll be included in the release notes + docs, and really it's up to users/trainers/etc to study these. I'd like to see some way of including "breaking changes" prominently in the release notes (@timlinux would this be possible?) for things like this (and the notes about 1.x project incompatibility).

Not sure what else we could/should do to make changes like this more noticeable to users. Maybe a "release notes" or "what's new" option in the help menu might help.

haubourg commented:
Hi Nyall,
your changes sound very good to me, thanks a lot for your work. I was a
little stressed out by the dry PR message and the fact that only my twitter
subscription on QGIS notifications drew my attention on that change.

@timlinux, I think we should recenter or link together our communication
channels, I see many people telling me it's impossible to follow QGIS
project since we don't have a unique channel for major changes. Some parts
are github only, other gitter only, and the dev mailing list does not
reflect all those topics.
Régis

2017-03-01 22:33 GMT+01:00 Nyall Dawson <notifications@github.com>:

> @hauborg I understand your concerns, but in the end this change won't be
> significantly different for users.
>
> I'll move the "no projection" option up from the tree view to its own
> widget to make it more prominent, and put it up near where the old OTF
> check box was.
>
> In the end the only really noticeable difference for users vs OTF-off is
> that measurements are all forced to be planar and unitless. (Which will
> hopefully finally put and end to the ongoing confusion about measurements
> in QGIS. Now, if you use a valid projection everything is perfect. If you
> use no projection, you get the only accurate measurements we can do
> (planar, no units)).
>
> So really from an end users pov this is just a string change (unchecking
> "on the fly" vs checking "no projection") + addition of the feature that
> it's now easy to make maps using a non earth based projection.
>
> It'll be included in the release notes + docs, and really it's up to
> users/trainers/etc to study these. I'd like to see some way of including
> "breaking changes" prominently in the release notes (@timlinux
> <https://github.com/timlinux> would this be possible?) for things like
> this (and the notes about 1.x project incompatibility).
>
> Not sure what else we could/should do to make changes like this more
> noticeable to users. Maybe a "release notes" or "what's new" option in the
> help menu might help.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/qgis/QGIS/pull/4198#issuecomment-283477287>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ABjenZrX-jQNCUsz9YeX7AwC5vJgwtkmks5rheQrgaJpZM4MN8TU>
> .
>

m-kuhn commented:
Hey, this broke my projects!!

I reopened a project today and it was all white and my map was gone, can you fix it please £àè!£

Actually, looking at the project in-depth showed, that

* in the (2.18 based) project CRS transform was turned off
* layers were in different CRS'es (21781 / 4326)
* features on the WGS 84 layer were on 6-digit coordinates

Main message for me: thanks for turning OTF on by default. It's a big win (because people will not as easily as in this case here mess up their data).

What I wonder though

* can we set the project CRS to invalid if OTF was off in a 2.18 project?
* can we make the "no projection" icon stand out with a red exclamation mark?

gioman commented:
> Main message for me: thanks for turning OTF on by default. It's a big win (because people will not as easily as in this case here mess up their data).

100% sure? ;) I'm not 100% sure, but anyway I haven't tried this change yet, I'll let you know my 2 cents, for what they are worth. Cheers!

m-kuhn commented:
100% sure!
It's still possible but the visibility of this (dangerous if not used consciously) mode is better.

saberraz commented:
I think it is best to file bugs to resolve those issues. See:

https://hub.qgis.org/issues/16338
https://hub.qgis.org/issues/16337
